src/Microsoft.AspNetCore.Mvc.Core/Internal/DefaultApplicationModelProvider.cs

```diff
@@ -213,24 +217,20 @@ protected virtual PropertyModel CreatePropertyModel(PropertyInfo propertyInfo)
                 throw new ArgumentNullException(nameof(propertyInfo));
             }
 
-            // CoreCLR returns IEnumerable<Attribute> from GetCustomAttributes - the OfType<object>
-            // is needed to so that the result of ToArray() is object
             var attributes = propertyInfo.GetCustomAttributes(inherit: true);
-            var propertyModel = new PropertyModel(propertyInfo, attributes);
-            var bindingInfo = BindingInfo.GetBindingInfo(attributes);
-            if (bindingInfo != null)
+            var propertyModel = new PropertyModel(propertyInfo, attributes)
             {
-                propertyModel.BindingInfo = bindingInfo;
-            }
-            else if (IsFormFileType(propertyInfo.PropertyType))
+                PropertyName = propertyInfo.Name
+            };
+
+            var modelMetadata = _modelMetadataProvider.GetMetadataForProperty(propertyInfo.DeclaringType, propertyInfo.Name);
+            var bindingInfo = BindingInfo.GetBindingInfo(modelMetadata) ?? BindingInfo.GetBindingInfo(attributes);
```

Comment:
There's a slight discrepancy in how `ModelMetadata` and `BindingInfo.GetBindingInfo` infer attributes.

For instance, `SomeAction([ModelMetadata] int value)`, `ModelMetadata.BinderType` / `ModelMetadata.BinderName` are null, but `BindingInfo.GetBindingInfo(attributes)` produces a non-null result (since all it does is look at the presence of the attribute). I couldn't see an obvious way to determine the presence of the attribute via `ModelMetadata`.

There are a few tests (particularly in the WebApiCompatShim space) that rely on the empty `ModelMetadata` -> no null `BindingInfo` behavior.

Reply:
`[ModelMetadata]` doesn't exist. What attribute do you mean? Will the discrepancy go away if you address my earlier comment and start with the `BindingInfo` that `BindingInfo.GetBindingInfo(...)` returns (if any)?

Reply:
Sorry, I meant `ModelBinderAttribute`. The discrepancy wouldn't go away. That said, we could stash an additional value on `ModelMetadata` that indicates the presence of any form of model binding attribute (equivalent to the  `isBindingInfoPresent` that's calculated in `BindingInfo`). If we have that information, we could remove the fall back.